feat(#9948): couchdb haproxy unit tests

shashi-sah2003 opened the pull request:
This draft pull request introduces the CouchDB unit testing setup.

<!-- DESCRIPTION -->

Closes #9948 

# Code review checklist
<!-- Remove or comment out any items that do not apply to this PR; in the remaining boxes, replace the [ ] with [x]. -->
- [ ] UI/UX backwards compatible: Test it works for the new design (enabled by default). And test it works in the old design, enable `can_view_old_navigation` permission to see the old design. Test it has appropriate design for RTL languages. 
- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

<!-- COMPOSE URLS GO HERE - DO NOT CHANGE -->

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

shashi-sah2003 commented:
@dianabarsan I am currently facing issue with cluster. According to current logs it is failing with tests related to cluster. 
I need your help with this on where I am making mistake.

```
mocha-test-1  |   CouchDB Cluster Unit Tests
mocha-test-1  | Setting up test database...
mocha-test-1  | Created test database 'tests'
mocha-test-1  |     ✔ couchdb1 node is available
mocha-test-1  |     ✔ couchdb2 node is available
mocha-test-1  |     ✔ couchdb3 node is available
mocha-test-1  | Testing document replication with id: kt3s6jtdkl
mocha-test-1  |     1) data inserted on one server can be retrieved from a peer
mocha-test-1  |     2) cluster membership shows all nodes
mocha-test-1  | Cleaning up test database...
mocha-test-1  | 
mocha-test-1  | 
mocha-test-1  |   3 passing (137ms)
mocha-test-1  |   2 failing
mocha-test-1  | 
mocha-test-1  |   1) CouchDB Cluster Unit Tests
mocha-test-1  |        data inserted on one server can be retrieved from a peer:
mocha-test-1  |      Error: Request failed with status 404
mocha-test-1  |       at requestOnNode (couch-cluster.spec.js:34:23)
mocha-test-1  |       at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
mocha-test-1  |       at async Context.<anonymous> (couch-cluster.spec.js:108:5)
mocha-test-1  | 
mocha-test-1  |   2) CouchDB Cluster Unit Tests
mocha-test-1  |        cluster membership shows all nodes:
mocha-test-1  | 
mocha-test-1  |       AssertionError: expected [ 'couchdb@couchdb1' ] to be a superset of [ 'couchdb@couchdb1', …(2) ]
mocha-test-1  |       + expected - actual
mocha-test-1  | 
mocha-test-1  |        [
mocha-test-1  |          "couchdb@couchdb1"
mocha-test-1  |       +  "couchdb@couchdb2"
mocha-test-1  |       +  "couchdb@couchdb3"
mocha-test-1  |        ]
mocha-test-1  |       
mocha-test-1  |       at Context.<anonymous> (couch-cluster.spec.js:126:45)
mocha-test-1  |       at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
mocha-test-1  | 
mocha-test-1  | 
mocha-test-1  | 
mocha-test-1 exited with code 2
```

dianabarsan commented:
> failing with tests related to cluster.

I haven't tried running your tests, but it can be that setting up the cluster takes a little time and you're starting the test suite _before_ that completes. So I would change the test so that there is a retry - I believe that the current suite has an initial test setup step that waits for the cluster to be completed: https://github.com/medic/cht-core/blob/master/couchdb/tests/tests.sh#L7

shashi-sah2003 commented:
@dianabarsan there was some underlying issue regarding domain name configuration so after setting it to `couchdb1.local`, `couchdb2.local`, `couchdb3.local`. so now all 5 unit test cases are passing for couchdb.

Also I have deleted some files which I thought was now redundant after this js test implementation. Could you please look at those files that if I might have deleted any important files maybe?

shashi-sah2003 commented:
@dianabarsan I have ported the haproxy unit test to js as well. but there is one issue I am facing with [should return json on connection drop](https://github.com/medic/cht-core/pull/9963/files#diff-5642fa4b76547d9582385a13a472007bc4e368faab8714b5bdec671f83440502:~:text=%7D)%3B-,it(%27should%20return%20json%20on%20connection%20drop%27%2C%20async,%7D)%3B,-3%20changes%3A%202)

the above test is not returing `502 Bad Gateway` which was defined in `with_mock.bats` file, it is returning the below error code when used with `fetch`.

```

Connection dropped, expected error: TypeError: fetch failed
    at node:internal/deps/undici/undici:13502:13
    at async Context.<anonymous> (/workspaces/cht-core/haproxy/tests/with_mock.spec.js:50:24) {
  [cause]: SocketError: other side closed
      at Socket.<anonymous> (node:internal/deps/undici/undici:6294:28)
      at Socket.emit (node:events:536:35)
      at endReadableNT (node:internal/streams/readable:1698:12)
      at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
    code: 'UND_ERR_SOCKET',
    socket: {
      localAddress: '127.0.0.1',
      localPort: 42334,
      remoteAddress: '127.0.0.1',
      remotePort: 15984,
      remoteFamily: 'IPv4',
      timeout: undefined,
      bytesWritten: 532,
      bytesRead: 584
    }
  }
}

```

Should I change the logic from `502 Bad Gateway` to `UND_ERR_SOCKET`?

dianabarsan commented on couchdb/tests/couch-cluster.spec.js:
```diff
@@ -0,0 +1,125 @@
+    
+    try {
+      const response = await fetch(url, fetchOptions);
+      const body = await response.json().catch(() => ({}));
```
I think that we should throw an error when the `.json` call fails.

dianabarsan commented on couchdb/tests/package.json:
```diff
@@ -0,0 +1,13 @@
+  "scripts": {
+    "test": "mocha couch-cluster.spec.js --no-config"
+  },
+  "dependencies": {
```
Can you please change this so the tests are run on the main machine, and not in a container? This will mean that the couchdbs will have to expose their ports (i normally use custom ports like: 15984, 25984, 35984). 
The main advantage will be that we will have one less container to create and install _and_ that we will not need to worry about upgrading or maintaining these dependencies, as the test will use the ones already installed on the main machine.

shashi-sah2003 replied:
@dianabarsan do you want me to remove the couchdb tests from running on container and shift it to local machine?

dianabarsan replied:
CouchDb should run in a container. The mocha engine should run on the main machine and test against the containerized CouchDb.

dianabarsan commented on haproxy/tests/errors_generate.spec.js:
```diff
@@ -0,0 +1,65 @@
+const { expect } = require('chai');
```
We typically use a `-`to separate words in file names. I know these existent files did not honor this convention. Would you mind renaming your newly added files with a `-` instead and make the world a little bit more consistent? :D

dianabarsan commented on haproxy/tests/package.json:
```diff
@@ -0,0 +1,16 @@
+    "test:errors": "mocha --timeout 10000 errors_generate.spec.js",
+    "test:integration": "mocha --timeout 30000 with_mock.spec.js"
+  },
+  "dependencies": {
+    "chai": "^4.3.7",
+    "chai-exclude": "^2.1.1",
+    "fs-extra": "^11.1.1",
+    "mocha": "^10.2.0",
+    "tmp-promise": "^3.0.3"
```
all these dependencies should be installed as dev dependencies in the main package json file.

dianabarsan commented on haproxy/tests/with_mock.spec.js:
```diff
@@ -0,0 +1,56 @@
+describe('HAProxy with mock CouchDB', function() {
+  this.timeout(3000);
+  
+  before(async () => {
+  });
+  
+  after(() => {
+  });
```
Please remove empty before and afters.

dianabarsan commented on haproxy/tests/with_mock.spec.js:
```diff
@@ -0,0 +1,56 @@
+      }
+    } catch (error) {
+      // Wait and try again
+      await new Promise(resolve => setTimeout(resolve, delay));
```
There's a bit of an inconsistency here, where we add a timeout when fetch throws an error but we don't add a timeout when the request fails (!response.ok). I think this should be refactored to have a timeout for both cases, 
Additionally, what is the purpose of adding the `delay` and `maxAttempts` as a parameter here? Neither seem to be used in any function call

shashi-sah2003 replied:
@dianabarsan You're right, the parameters weren't being customized in the test cases, but I kept them to make the function reusable in case we want to adjust retry behavior later. The defaults handle most use cases, but having them as parameters gives us flexibility if needed.

dianabarsan commented on haproxy/tests/with_mock.spec.js:
```diff
@@ -0,0 +1,56 @@
+    try {
+      const response = await fetch(url, { 
+        method: 'GET',
+        timeout: 5000
```
afaik Request doesn't have a timeout property: https://developer.mozilla.org/en-US/docs/Web/API/Request
You need to use an AbortSignal for this. 

Also, I think the default method will be GET. You don't need to specify it.

dianabarsan reviewed (changes requested):
Really excellent work, I am so impressed! These changes are wonderful and needed improvements to our codebase, make it more consistent and maintainable and approachable. Thank you so much! 

I left some comments, but nothing major. 
This PR made me very happy!

shashi-sah2003 commented:
> Really excellent work, I am so impressed! These changes are wonderful and needed improvements to our codebase, make it more consistent and maintainable and approachable. Thank you so much!
> 
> I left some comments, but nothing major. This PR made me very happy!

@dianabarsan thank you so much. I have made the all the changes except one for which I have asked you a doubt. and also please give me your insights on this [issue](https://github.com/medic/cht-core/pull/9963#issuecomment-2883150384).

dianabarsan reviewed (commented):
As above (replied in thread). 
The `couch-test` container should be removed and requests should be made against the couchdb containers.

shashi-sah2003 commented:
@dianabarsan I have made the requested changes to remove `couch-test` containers and run mocha engine locally. Also as I have mentioned earlier what should be done for this test failing `should return json on connection drop` in 

> @dianabarsan I have ported the haproxy unit test to js as well. but there is one issue I am facing with [should return json on connection drop](https://github.com/medic/cht-core/pull/9963/files#diff-5642fa4b76547d9582385a13a472007bc4e368faab8714b5bdec671f83440502)
> 
> the above test is not returing `502 Bad Gateway` which was defined in `with_mock.bats` file, it is returning the below error code when used with `fetch`.
> 
> ```
> 
> Connection dropped, expected error: TypeError: fetch failed
>     at node:internal/deps/undici/undici:13502:13
>     at async Context.<anonymous> (/workspaces/cht-core/haproxy/tests/with_mock.spec.js:50:24) {
>   [cause]: SocketError: other side closed
>       at Socket.<anonymous> (node:internal/deps/undici/undici:6294:28)
>       at Socket.emit (node:events:536:35)
>       at endReadableNT (node:internal/streams/readable:1698:12)
>       at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
>     code: 'UND_ERR_SOCKET',
>     socket: {
>       localAddress: '127.0.0.1',
>       localPort: 42334,
>       remoteAddress: '127.0.0.1',
>       remotePort: 15984,
>       remoteFamily: 'IPv4',
>       timeout: undefined,
>       bytesWritten: 532,
>       bytesRead: 584
>     }
>   }
> }
> ```
> 
> Should I change the logic from `502 Bad Gateway` to `UND_ERR_SOCKET`?

@dianabarsan I have made the requested changes to remove `couch-test` containers and run mocha engine locally. Also as I have mentioned earlier what should be done for this test failing `should return json on connection drop` in `with-mock.spec.js`. The `fetch` command is not giving expected output as `curl` which was used earlier for connection drop.

dianabarsan commented on couchdb/tests/package.json:
```diff
@@ -0,0 +1,13 @@
+  "scripts": {
+    "test": "npx mocha couch-cluster.spec.js --no-config"
+  },
+  "dependencies": {
+    "chai": "^4.3.8",
+    "chai-exclude": "^2.1.1",
+    "mocha": "^11.1.0"
+  }
```
These should not be needed. Please remove.

dianabarsan commented on couchdb/tests/Makefile:
```diff
@@ -1,15 +1,29 @@
+	docker compose down -v --remove-orphans
+	docker compose up -d couchdb1 couchdb2 couchdb3
+	$(MAKE) wait-for-couchdb
+	npm install --no-save
```
npm install should not be run. all dependencies should have already been installed in through the root env.

dianabarsan commented on couchdb/tests/couch-cluster.spec.js:
```diff
@@ -0,0 +1,127 @@
+
+  after(async function () {
+    
+    if (dbCreated) {
```
Do we have to cleanup? Ww destroy the containers after the test anyway.

shashi-sah2003 replied:
@dianabarsan you're right. This is redundant hook. Thanks for pointing this out also.

dianabarsan commented on couchdb/tests/couch-cluster.spec.js:
```diff
@@ -0,0 +1,127 @@
+  it('should create the test database', async function () {
+    const res = await requestOnNode(0, `/${TEST_DB_NAME}`, { method: 'PUT'});
+
+    expect([201, 202, 412]).to.include(res.status);
```
How come `412` is an accepted status here?

dianabarsan commented on haproxy/tests/Makefile:
```diff
@@ -1,16 +1,21 @@
 .PHONY: clean
 clean:
 	docker compose down --rmi local || :
+	rm -rf js/node_modules
```
When does this folder get created?

shashi-sah2003 replied:
sorry @dianabarsan earlier I was working inside js directory and forgot to remove this

dianabarsan commented on haproxy/tests/package.json:
```diff
@@ -0,0 +1,16 @@
+    "test:errors": "mocha --timeout 10000 errors-generate.spec.js",
+    "test:integration": "mocha --timeout 30000 with-mock.spec.js"
+  },
+  "devDependencies": {
+    "chai": "^4.3.7",
+    "chai-exclude": "^2.1.1",
+    "fs-extra": "^11.1.1",
+    "mocha": "^10.2.0",
+    "tmp-promise": "^3.0.3"
+  }
```
These should not be needed. Please remove.

dianabarsan commented on couchdb/tests/Makefile:
```diff
@@ -1,15 +1,29 @@
+wait-for-couchdb:
+	@echo "Waiting for CouchDB containers to be ready..."
+	@for i in 1 2 3; do \
+		until curl -s -f -u "$(COUCHDB_USER):$(COUCHDB_PASSWORD)" http://localhost:$${i}5984/_up > /dev/null; do \
+			echo "Waiting for couchdb$$i..."; \
```
I like this 1 2 3 iteration :D

dianabarsan reviewed (changes requested):
Lovely! Thanks for making the tests work on the base machine!
I left a few comments below.

dianabarsan reviewed (commented):
Hi @shashi-sah2003 

I believe you're running into some linting issues https://github.com/medic/cht-core/actions/runs/15187911182/job/42712846711?pr=9963
Additionally, there is one CI job that now only runs conditionally When changes are made to CouchDb: https://github.com/medic/cht-core/blob/master/.github/workflows/test_couchdb.yml
This workflow is failing now because we don't run the global npm ci, but I believe that we don't need this workflow now since we're running the CouchDb tests every time. So this workflow file can be safely deleted.

shashi-sah2003 commented:
@dianabarsan I have fixed the linting issue and removed the test_couchdb.yml file. 
https://github.com/medic/cht-core/actions/runs/15190362944/job/42721312557?pr=9963 this is the current issue we are facing right now. I have also mentioned this [here](https://github.com/medic/cht-core/pull/9963#issuecomment-2899991765)

dianabarsan reviewed (commented):
That error is pretty weird. It seems like other people are getting it in various situations: https://github.com/nodejs/undici/issues/3492
It could be this bug in `fetch` that you are seeing. 

I suppose the solution is to not use `fetch`, but to `exec` the `curl` command within JS: 
```
const util = require('node:util');
const exec = util.promisify(require('node:child_process').exec);

const { stdout, stderr } = await exec('curl -s http://127.0.0.1:15984/error/drop');
```

and then run the assertions over the returned values. We could add a comment about the fetch issue.

shashi-sah2003 commented:
@dianabarsan I have made the requested changes. There was some issue with running couchdb test on GHA but were working fine locally. For that I have added another logic in the makefile to wait for cluster formation before proceeding now that is also passing. I think this PR is ready to merge

shashi-sah2003 commented on couchdb/tests/.env:
```diff
@@ -0,0 +1,2 @@
+COUCHDB_PASSWORD=pass
+COUCHDB_USER=medic-test-admin
```
@dianabarsan when running docker command directly from `package.json` earlier the `.env` file was necessary but now we are running `make` command directly from `package.json` file so env file isn't necessary

shashi-sah2003 commented on haproxy/tests/with-mock.spec.js:
```diff
@@ -0,0 +1,56 @@
+      if (response.ok) {
+        return true;
+      }
+      await new Promise(resolve => setTimeout(resolve, delay));
```
yeah it will work

shashi-sah2003 commented on couchdb/tests/Makefile:
```diff
@@ -1,15 +1,34 @@
 BACKSLASH := \$ #
-COUCHDB_PASSWORD := test_~!@\#$$%^&*_-+=`|$(BACKSLASH)(){}[]:;"'<>,.?/_pw
+COUCHDB_PASSWORD := pass
```
@dianabarsan The issue with using `test_~!@\#$$%^&*_-+=`|$(BACKSLASH)(){}[]:;"'<>,.?/_pw` is with `backtick` with is causing this error.
```/bin/sh: 6: Syntax error: EOF in backquote substitution
```
after removing `backtick`, the issue was with `'` and `"` so I have also removed these as these were conflicting with this line w/o termination error 
```until curl -s -f -u "$(COUCHDB_USER):$(COUCHDB_PASSWORD)" http://localhost:$${i}5984/_up > /dev/null; do \```

pls let me know if the removed symbols don't make much effect or if it is imperative to keep them and is there any other alternative for this.

shashi-sah2003 commented:
@dianabarsan I have made the requested changes above and asked for a suggestion in the component of code above. Do let me know what do you think about that?

shashi-sah2003 commented on couchdb/Dockerfile:
```diff
@@ -24,7 +24,6 @@ FROM alpine:3 AS test_couchdb_build
 RUN apk --no-cache add bash bats curl jq openssl
 COPY tests /app/tests
```
@dianabarsan  
No, I have checked the without it, it is working fine
thanks for noticing :)

shashi-sah2003 commented:
> Sweet! I love this :)
> 
> I've restarted the two jobs that failed and this should be ready to be merged after success. Great work!

@dianabarsan I have removed that section of the code which was not necessary. There are some failing jobs, Could you pls rerun from your end?

dianabarsan reviewed (approved):
I love it! This makes the testing so much more approachable and maintainable! Thank you so much for this contribution!
